Revert immutable varargs

szeiger opened the pull request:
I have no idea why it won't bootstrap...

Fixes https://github.com/scala/scala-dev/issues/591

szeiger committed:
Revert "Copy arrays that are explicitly passed as varargs with `: _*`"

This reverts commit 05117dcd5f88e61327f3b27591453599dfe20015.

szeiger committed:
Use scala.collection.Seq for varargs

This reverts parts of 7c018a0056e4ee0f6f70fcfcabc00fa440eeb83b.

- `SeqModule` in `Definitions` continues to refer to `s.c.i.Seq`
  because it is expected to match `scala.Seq`. The definition of
  `SeqClass`, which is only used for varargs handling, has been reverted
  to refer to `s.c.Seq`.

- Update `repeatedToSeq` in `Unapplies` because it relied on varargs
  being expressed as `scala.Seq`

- Remove the conditional support for pre-M4 collections.

- Update the spec to refer to `scala.collection.Seq` rather than `Seq`
  or `scala.Seq` for repeated parameters / varargs and `unapplySeq`
  extractors.

- Add the required `toSeq` and other calls to convert from varargs to
  scala.Seq.

- Add a check for immutable varargs in the standard library to the
  compiler (in `Global`) to allow bootstrapping with sbt.

adriaanm commented:
the exception is truncated, but there are some hints that could lead to colonel Array in the library with the varargs?
```
scala.Array$.apply(Lscala/collection/Seq;Lscala/reflect/ClassTag;)Ljava/lang/Object;
xsbt.ExtractAPI$Constants$.<init>(ExtractAPI.scala:736)
```

szeiger commented:
The exception indicates that code compiled with the bootstrapped compiler tries to call into `Array.apply` in the non-bootstrapped library, which should not happen and which did not happen when making the step in the opposite direction a few months ago.

You can see this error when using the non-bootstrapped compiler (by running `sbt scala` in the publish-core build):

```
scala> Array(new scala.collection.mutable.ArrayBuffer[Int]: _*)
java.lang.NoSuchMethodError: scala.Array$.apply(Lscala/collection/Seq;Lscala/reflect/ClassTag;)Ljava/lang/Object;
  ... 29 elided
```

This is expected because the library was still compiled for immutable varargs and the call-site is compiled for generic varargs.

So let's try the same with some freshly compiled code:

```
scala> object Foo { def apply[T](xs: T*): String = xs.getClass.getName }
defined object Foo

scala> :javap Foo
[...]
  public <T extends java.lang.Object> java.lang.String apply(scala.collection.Seq<T>);
[...]

scala> Foo(new scala.collection.mutable.ArrayBuffer[Int]: _*)
res7: String = scala.collection.mutable.ArrayBuffer
````

This works as expected. The code compiled with the new compiler uses `scala.collection.Seq` for the varargs method and the method can be successfully called.

In the bootstrapped build the compiler and library were still compiled with immutable varargs but the compiler will produce code for generic varargs. The first thing this compiler compiles is the standard library and all new code then uses that standard library (with generic varargs). But this first step (compiling the standard library) is the one that fails which could indicate that it has access to the old standard library (which it should not have).

szeiger commented:
This is the call site: https://github.com/sbt/zinc/blob/develop/internal/compiler-bridge/src/main/scala/xsbt/API.scala. I think we're running into a fundamental problem with bootstrapping here. I wonder why this was not the case when doing the original switch to immutable varargs. Was this done differently in sbt 0.13 (which I think we still used at the time)?

adriaanm commented:
It sounds very plausible that the compilation of the compiler interface is the problem. Could we try on sbt 0.13 to rule that out?

szeiger commented:
I've discussed the problem with @eed3si9n. This is a limitation of Zinc that could be fixed. It currently does not distinguish between the compiler to use for compiling the bridge and the compiler to compile against so it only works correctly for fully bootstrapped Scala versions (where the two are identical). It would be nice to get this fixed eventually because it would make bootstrapping easier in the long run.

For now I've added a hack to determine immutable vs generic varargs handling so that we can bootstrap successfully.

szeiger commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -56,6 +56,21 @@ class Global(var currentSettings: Settings, reporter0: LegacyReporter)
+  //TODO remove after bootstrapping
+  // sbt uses this compiler to compile its own compiler bridge, therefore we need to use immutable varargs for
+  // compatibility if the library uses them
+  override lazy val isImmutableVarargs: Boolean = try {
```
This is quite the hack and it logs a few errors while running the tests but it works well enough to allow bootstrapping.

szeiger commented on src/reflect/scala/reflect/internal/Trees.scala:
```diff
@@ -1552,9 +1552,9 @@ trait Trees extends api.Trees {
-      case Seq(_, rest @ _*) => Block(stats.init.toList, stats.last)
+      case collection.Seq(b @ Block(_, _)) => b
+      case collection.Seq(stat) => Block(stats.toList, Literal(Constant(())))
+      case collection.Seq(_, rest @ _*) => Block(stats.init.toList, stats.last)
```
Gotcha: Varags are not Seqs anymore. This was the ugliest side-effect that I encountered. The code still compiled but some tests failed.

dwijnand replied:
Why can't we let Seq.unapply handle scala.collection.Seqs?

szeiger replied:
Why should `immutable.Seq` match a `mutable.Seq`? It would be totally inconsistent.

eed3si9n replied:
This is tricky. Many people have gotten used to the following behavior:

```scala
scala> val xs = scala.collection.mutable.ListBuffer(1, 2, 3)
xs: scala.collection.mutable.ListBuffer[Int] = ListBuffer(1, 2, 3)

scala> xs match { case Seq(xs @ _*) => xs }
res0: Seq[Int] = ListBuffer(1, 2, 3)
```

Technical reason may have been that `scala.Seq` was `scala.collection.Seq` etc, but this source compatibility break is subtle, so I think it's worth being hand-holding about `2.[0 - 12]` code.

dwijnand replied:
I'm thinking https://en.wikipedia.org/wiki/Robustness_principle.

Having `scala.Seq` be different to `scala.collection.Seq` is very confusing.  The fact that Scala's pattern matching isn't exhaustive further aggravates the situation.  Could we not introduce `scala.Varargs` instead?

szeiger commented:
What were the chances of getting a merge conflict in that one line in `Unapplies`? Everything should be green now (unless I messed up serialization stability while rebasing).

szeiger commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -56,6 +56,21 @@ class Global(var currentSettings: Settings, reporter0: LegacyReporter)
+    //Console.err.println("**** varargs class is: "+varargsClass)
+    varargsClass == "scala.collection.immutable.Seq"
+  } catch { case t: Throwable =>
+    //Console.err.println("Error determining `isImmutableVarargs`, assuming `false`: "+t)
```
I had to comment this out because it messed up hundreds of partests. The code appears to work and return `true` in the one place where it needs to (when compiling sbt's compiler bridge). `false` is the correct result in any other situation.

szeiger commented on src/reflect/scala/reflect/runtime/JavaMirrors.scala:
```diff
@@ -478,14 +478,14 @@ private[scala] trait JavaMirrors extends internal.SymbolTable with api.JavaUnive
         def objReceiver       = receiver.asInstanceOf[AnyRef]
         def objArg0           = args(0).asInstanceOf[AnyRef]
-        def objArgs           = args.asInstanceOf[Seq[AnyRef]]
+        def objArgs           = args.asInstanceOf[scala.collection.Seq[AnyRef]]
```
Some even worse gotchas showed up here in `JavaMirrors`. I hope our tests are sufficiently complete to catch all of them. This is what you get when varargs are no longer Seqs.

adriaanm commented on src/library/scala/StringContext.scala:
```diff
@@ -91,7 +91,7 @@ case class StringContext(parts: String*) {
    *          use of a StringBuilder.
    */
-  def s(args: Any*): String = standardInterpolator(processEscapes, args, parts)
+  def s(args: Any*): String = standardInterpolator(processEscapes, args, parts.toSeq)
```
ah right, this method is now a fast-tracked macro, although it was already special-cased in the compiler (previously in refchecks)

adriaanm replied:
that's probably a red herring, though -- I think the real problem is in `caseModuleUnapplyMeth`

adriaanm replied:
this makes it bootstrap:
```diff
diff --git i/src/compiler/scala/tools/nsc/typechecker/Unapplies.scala w/src/compiler/scala/tools/nsc/typechecker/Unapplies.scala
index 5b3a8ff..a8a7559 100644
--- i/src/compiler/scala/tools/nsc/typechecker/Unapplies.scala
+++ w/src/compiler/scala/tools/nsc/typechecker/Unapplies.scala
@@ -164,7 +164,7 @@ trait Unapplies extends ast.TreeDSL {
         } getOrElse Select(unapplyParam, caseAccessorName(cdef.symbol, constrParam.name))
 
       constrParam.tpt match {
-        case AppliedTypeTree(Select(_, tpnme.REPEATED_PARAM_CLASS_NAME), tps) => Typed(accSel, AppliedTypeTree(gen.rootScalaDot(tpnme.Seq), tps))
+        case AppliedTypeTree(Select(_, tpnme.REPEATED_PARAM_CLASS_NAME), tps) => Select(accSel, nme.toSeq)
         case _                                                                => accSel
       }
     }
```

szeiger commented:
It looks like our built-in "quick" compiler/REPL also mixes up the classpaths:
```
$ sbt
[...]
sbt:root> scalac Test.scala
[...]
[info] Running (fork) scala.tools.nsc.Main -usejavacp Test.scala
**** varargs class is: scala.collection.immutable.Seq
[success] Total time: 157 s, completed Feb 5, 2019 12:30:10 PM
```
It should have immutable.Seq varargs on the JVM classpath but collection.Seq varargs on the compiler classpath.

adriaanm commented:
I added a comment above to explain & fix the issue

szeiger commented:
/rebuild

adriaanm commented:
/rebuild

szeiger commented:
More bootstrapping classpath issues in scalatest. After fixing the trivial compile error:

```
sbt:root> scalacheck/test
[info] Compiling 20 Scala sources to /Users/szeiger/code/scala/build/quick/classes/scalacheckLib ...
[warn] there were 57 deprecation warnings (since 2.13.0); re-run with -deprecation for details
[warn] one warning found
[info] Done compiling.
[info] Compiling 67 Scala sources to /Users/szeiger/code/scala/target/scalacheck/test-classes ...
[warn] there were three deprecation warnings (since 2.10.0)
[warn] there were 9 deprecation warnings (since 2.10.1)
[warn] there were 22 deprecation warnings (since 2.11.0)
[warn] there were 88 deprecation warnings (since 2.13.0)
[warn] there were 122 deprecation warnings in total; re-run with -deprecation for details
[warn] there were 13 feature warnings; re-run with -feature for details
[warn] 6 warnings found
[info] Done compiling.
[error] java.lang.NoSuchMethodError: scala.collection.immutable.Set$.apply(Lscala/collection/Seq;)Ljava/lang/Object;
[error] 	at org.scalacheck.Test$$anon$4$OptMinSuccess$.<init>(Test.scala:231)
[error] 	at org.scalacheck.Test$$anon$4.OptMinSuccess$lzycompute$1(Test.scala:229)
[error] 	at org.scalacheck.Test$$anon$4.OptMinSuccess(Test.scala:229)
[error] 	at org.scalacheck.Test$$anon$4.<init>(Test.scala:270)
[error] 	at org.scalacheck.Test$.cmdLineParser$lzycompute(Test.scala:228)
[error] 	at org.scalacheck.Test$.cmdLineParser(Test.scala:228)
[error] 	at org.scalacheck.CustomScalaCheckFramework$$anon$5.<init>(CustomScalaCheckRunner.scala:186)
[error] 	at org.scalacheck.CustomScalaCheckFramework.runner(CustomScalaCheckRunner.scala:181)
[error] 	at sbt.Defaults$.$anonfun$createTestRunners$1(Defaults.scala:944)
[error] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233)
[error] 	at scala.collection.immutable.Map$Map1.foreach(Map.scala:125)
[error] 	at scala.collection.TraversableLike.map(TraversableLike.scala:233)
[error] 	at scala.collection.TraversableLike.map$(TraversableLike.scala:226)
[error] 	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
[error] 	at sbt.Defaults$.createTestRunners(Defaults.scala:938)
[error] 	at sbt.Defaults$.allTestGroupsTask(Defaults.scala:1000)
[error] 	at sbt.Defaults$.$anonfun$testTasks$10(Defaults.scala:756)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:40)
[error] 	at sbt.std.Transform$$anon$4.work(System.scala:67)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:269)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error] 	at sbt.Execute.work(Execute.scala:278)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:269)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:178)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:37)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] 	at java.lang.Thread.run(Thread.java:748)
[error] (scalacheck / Test / executeTests) java.lang.NoSuchMethodError: scala.collection.immutable.Set$.apply(Lscala/collection/Seq;)Ljava/lang/Object;
```

The library looks OK:
```
$ javap -classpath build/quick/classes/library/ 'scala.collection.immutable.Set$'
Compiled from "Set.scala"
public final class scala.collection.immutable.Set$ implements scala.collection.IterableFactory<scala.collection.immutable.Set> {
  public static final scala.collection.immutable.Set$ MODULE$;
  public static {};
  public java.lang.Object apply(scala.collection.Seq);
```

The classpath that it *should* use looks good, too, so that's probably not the one it actually uses:
```
sbt:root> show scalacheck/test:fullClasspath
[info] * Attributed(/Users/szeiger/code/scala/target/scalacheck/test-classes)
[info] * Attributed(/Users/szeiger/code/scala/build/quick/classes/scalacheck)
[info] * Attributed(/Users/szeiger/code/scala/build/quick/classes/library)
[info] * Attributed(/Users/szeiger/code/scala/build/quick/classes/reflect)
[info] * Attributed(/Users/szeiger/code/scala/build/quick/classes/compiler)
[info] * Attributed(/Users/szeiger/code/scala/build/quick/classes/scaladoc)
[info] * Attributed(/Users/szeiger/code/scala/build/quick/classes/scalacheckLib)
[info] * Attributed(/Users/szeiger/.ivy2/cache/org.scala-lang.modules/scala-asm/bundles/scala-asm-7.0.0-scala-1.jar)
[info] * Attributed(/Users/szeiger/.ivy2/cache/jline/jline/jars/jline-2.14.6.jar)
[info] * Attributed(/Users/szeiger/.ivy2/cache/org.scala-sbt/test-interface/jars/test-interface-1.0.jar)
```

szeiger commented:
I left a comment on https://github.com/sbt/sbt/issues/4009#issuecomment-460736281. This looks like another sbt issue. I don't know why it doesn't affect the JUnit test, too.

szeiger committed:
Use lazy Framework implementation for ScalaCheck

This avoids issues when running on the wrong classpath due to
https://github.com/sbt/sbt/issues/4009.

szeiger commented:
Almost there! (with a failed Travis run because of a new incompatibility that was merged in the meantime)

adriaanm commented:
whew! 😅

szeiger commented:
"I just have to revert that one commit from the original change. Shouldn't take more than an hour or so." (Me, 7 days ago)

dwijnand commented on spec/08-pattern-matching.md:
```diff
@@ -266,7 +266,7 @@ the following applies:
 An `unapplySeq` method in an object $x$ matches the pattern
 $x(q_1 , \ldots , q_m, p_1 , \ldots , p_n)$ if it takes exactly one argument
-and its result type is of the form `Option[($T_1 , \ldots , T_m$, Seq[S])]` (if `m = 0`, the type `Option[Seq[S]]` is also accepted).
+and its result type is of the form `Option[($T_1 , \ldots , T_m$, scala.collection.Seq[S])]` (if `m = 0`, the type `Option[scala.collection.Seq[S]]` is also accepted).
```
I agree with @jrudolph in https://github.com/scala/scala-dev/issues/591#issuecomment-458895442, `_*` and `unapplySeq` should return Seqs, not `scala.collection.Seq`, which probably requires `Seq.unapply` to accept `scala.collection.Seq`s.

szeiger replied:
Thanks to name-based pattern matching you can already use different return types for `unapplySeq`. They are only converted to `scala.collection.Seq` (previously `scala.Seq`) when you match with `@ _*`. This is the dual of varargs, using the same `_*` syntax. It would be very inconsistent to give it a different type.

Note that this is unrelated to actually matching a Seq type (`case Seq(x, y, z) => ...`).

dwijnand replied:
One of the reasons for moving varargs back to `scala.collection.Seq` is performance.  Is that relevant for `..., xs: _*)` unapply too?

adriaanm commented:
I also think that we should try hard to keep the illusion that varargs are Seqs. Maybe we can have a varargs extractor that we recommend for future proofing, so that we can later switch these things around independently

adriaanm commented:
/cc @odersky & @smarter — let’s discuss on Monday

dwijnand commented:
Further, could varargs be its own type which cheaply toSeq's?  That would be one step further on making the Scala kernel independent of scala.collection.

szeiger commented:
An extractor won't fix the problem where people expect a scala.Seq from varargs in a pattern match. It may turn an exhaustive match into a non-exhaustive one at best. A different varargs type would be better (because it is a compiler error when the type is incompatible) and this is the plan for Dotty.

When it was decided that we should investigate going back to mutable varargs (which resulted in this PR) I thought that it was already too late for that. And even more so it is too late for a completely new solution with non-Seq varargs. There are too many angles to consider (interaction with scala.Seq; extractors (including the switch to name-based extractors); Java interop; JavaScript interop; etc.) to justify a rushed solution.

My preference is to keep immutable.Seq varargs (like in M4) for 2.13. A proper solution with a non-Seq varargs type needs more time and, while switching to immutable varargs now may create a bit more work in total for downstream projects (collection.Seq varargs -> immutable.Seq varargs -> separate varargs type or immutable arrays), there are fewer surprises where code compiles but fails at runtime.

adriaanm commented:
First, I agree we shouldn't rush anything -- I'm just looking for small refinements to ease the pain of the mismatch between the new default `Seq` and the varargs's idea of a `Seq`.

> An extractor won't fix the problem where people expect a scala.Seq from varargs in a pattern match.

True, but it would set us on the right path, and it would simplify migrating from 2.13 to 2.14, assuming we already implement the new varargs there (which we could, since we should hopefully also have opaque types)

szeiger commented:
The easiest way to avoid the mismatch is to keep using `scala.Seq` for varargs. The arguments that have been brought forth against immutable varargs are not very convincing IMHO:

- Clashes in erased signatures: Annoying when it happens but it is a corner case with decent workarounds (which may require different sources when cross-building). I've seen this with other changes in 2.13 (like clever overloads with `Traversable[Foo]` and `Iterable[Bar]`) and it will happen again when making varargs immutable arrays.

- Performance: You have to copy existing mutable Seqs to make them immutable.
  - A future solution based on immutable arrays will require copying anyway. But immutable arrays (and `immutable.ArraySeq` which we use in M4) already offer us a solution: Build a mutable array and wrap it with `unsafeWrapArray`.
  - You can use a wrapper that makes a mutable Seq look immutable. But you probably don't want to: When calling a Java method you still have to copy. Wrapping a mutable array avoids this. The Java bridge methods can wrap and unwrap as required.

dwijnand commented:
Wrapping makes it unmodifiable (not immutable) though, which I don't think is what scala.collection.immutable.Seq should be.

adriaanm commented:
> unmodifiable (not immutable)

I don't understand the difference. As long as the underlying mutable array is not aliased, that's an acceptable implementation of an immutable seq, no? Since we don't have language support for managing aliasing, it's on the programmer to ensure exclusive access to the array to be wrapped.

adriaanm commented:
Based on this discussion, I'm now leaning against merging this PR and keeping things as they currently stand (as was my initial position). When we change varargs again in 2.14/3.0, there will be some migration pain regardless of whether we merge this PR now, and it seems the delta in future pain is pretty small between both, whereas merging this PR seems to give us a more painful migration from 2.12/2.13.0-M5.

dwijnand commented:
> it's on the programmer to ensure exclusive access to the array to be wrapped

The problem is what does the varargs receiving programmer do?  Defensively copy because the `scala.Seq` it got it's _actually_ immutable?

That's the problem @xuwei-k pointed out in https://github.com/scala/bug/issues/11024#issuecomment-405508448, which is reverted with the first commit in the PR.

adriaanm commented:
I would say the bug is in the caller who didn't respect exclusive access to the mutable array. Maybe we can add something like https://swift.org/blog/swift-5-exclusivity/ to the Scala 3 feature list :-D

adriaanm commented:
(Statically, of course -- we wouldn't do a run-time check..)

szeiger commented:
https://github.com/scala/bug/issues/11024 was solved by making these unsafe conversions explicit. I think it's reasonable to expect that a varargs parameter is effectively immutable. Even though Java uses simple arrays I have never seen them used in a different way.

dwijnand commented:
"effectively immutable" isn't great.  But I've made my argument by now.

adriaanm commented:
Discussed with Martin and Guillaume and co. Agreed on keeping varargs as immutable Seqs
